src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs

```diff
@@ -104,40 +109,35 @@ private static void ClearCachedData()
             var (patternName, patternContainer) = PatternMatcher.GetNameAndContainer(searchPattern);
             var declaredSymbolInfoKindsSet = new DeclaredSymbolInfoKindSet(kinds);
 
-            await SearchCachedDocumentsInCurrentProcessAsync(
+            await foreach (var item in SearchCachedDocumentsInCurrentProcessAsync(
                 storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
-                onItemFound, priorityDocumentKeys, cancellationToken).ConfigureAwait(false);
+                priorityDocumentKeys, cancellationToken).WithCancellation(cancellationToken))
+            {
+                yield return item;
+            }
 
-            await SearchCachedDocumentsInCurrentProcessAsync(
+            await foreach (var item in SearchCachedDocumentsInCurrentProcessAsync(
                 storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
-                onItemFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+                lowPriDocs, cancellationToken).WithCancellation(cancellationToken))
+            {
+                yield return item;
+            }
         }
 
-        private static async Task SearchCachedDocumentsInCurrentProcessAsync(
+        private static IAsyncEnumerable<RoslynNavigateToItem> SearchCachedDocumentsInCurrentProcessAsync(
             IChecksummedPersistentStorageService storageService,
             string patternName,
             string? patternContainer,
             DeclaredSymbolInfoKindSet kinds,
-            Func<RoslynNavigateToItem, Task> onItemFound,
             ImmutableArray<DocumentKey> documentKeys,
             CancellationToken cancellationToken)
         {
-            using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+            using var _ = ArrayBuilder<IAsyncEnumerable<RoslynNavigateToItem>>.GetInstance(out var builder);
 
             foreach (var documentKey in documentKeys)
-            {
-                tasks.Add(Task.Run(async () =>
-                {
-                    var index = await GetIndexAsync(storageService, documentKey, cancellationToken).ConfigureAwait(false);
-                    if (index == null)
-                        return;
-
-                    await ProcessIndexAsync(
-                        documentKey.Id, document: null, patternName, patternContainer, kinds, onItemFound, index, cancellationToken).ConfigureAwait(false);
-                }, cancellationToken));
-            }
+                builder.Add(ProcessStaleIndexAsync(storageService, patternName, patternContainer, kinds, documentKey, cancellationToken));
 
-            await Task.WhenAll(tasks).ConfigureAwait(false);
+            return builder.ToArray().MergeAsync(cancellationToken);
```

Comment:
MergeAsync (which you'll see later in the PR) allows us to take `N` IAsyncEnumerables and combine them into a single final stream.  the individual values can all be interleaved.